src/memray/_memray.pyx

```diff
@@ -335,21 +341,106 @@ def start_thread_trace(frame, event, arg):
     return start_thread_trace
 
 
+cdef class ProgressIndicator:
+    cdef bool _report_progress
+    cdef object _indicator
+    cdef object _context_manager
+    cdef object _task
+    cdef object _task_description
+    cdef object _total
+    cdef size_t _cumulative_num_processed
+    cdef size_t _update_interval
+    cdef size_t _ns_between_refreshes
+    cdef timespec _next_refresh
+
+    def __init__(self,
+        str task_description,
+        object total,
+        bool report_progress=True,
+        size_t refresh_per_second=10,
+    ):
+        self._report_progress = report_progress
+        self._total = total
+        self._cumulative_num_processed = 0
+        # Only check the elapsed time every N records
+        self._update_interval = 100_000
+        self._ns_between_refreshes = 1_000_000_000 // refresh_per_second
+        self._next_refresh.tv_sec = 0
+        self._next_refresh.tv_nsec = 0
+        self._task_description = task_description
+        self._task = None
+        self._context_manager = None
+        if report_progress:
+            self._indicator = Progress(
+                auto_refresh=False,
+                transient=True,
+            )
+
+    def __enter__(self):
+        if not self._report_progress:
+            return self
+        self._context_manager = self._indicator.__enter__()
+        self._task = self._context_manager.add_task(
+            f"[blue]{self._task_description}...",
+            total=self._total
+        )
+        return self
+
+    def __exit__(self, type, value, traceback):
+        if not self._report_progress:
+            return
+        return self._context_manager.__exit__(type, value, traceback)
+
+    cdef bool _time_for_refresh(self):
+        cdef timespec now
+        cdef int rc = clock_gettime(CLOCK_MONOTONIC, &now)
+
+        if 0 != rc:
+            return True
+
+        if now.tv_sec > self._next_refresh.tv_sec or (
+            now.tv_sec == self._next_refresh.tv_sec
+            and now.tv_nsec > self._next_refresh.tv_nsec
+        ):
+            self._next_refresh = now
+            self._next_refresh.tv_nsec += self._ns_between_refreshes
+            while self._next_refresh.tv_nsec > 1_000_000_000:
+                self._next_refresh.tv_nsec -= 1_000_000_000
+                self._next_refresh.tv_sec += 1
+            return True
+
+        return False
+
+    cdef void update(self, size_t n_processed) except*:
```

Comment:
Why make this `cdef void` with an `except*` instead of just making it
```suggestion
    cdef update(self, size_t n_processed):
```
? I'd expect `except *` to be slightly slower, since that requires an extra call to `PyErr_Occurred()` per call, instead of just being able to compare a pointer against `NULL`. Some rough profiling seems to bear out that it's very slightly slower...

Reply:
I added that as a temporary fix to the exceptions being raised and I forgot to clean it up. Good catch!